src/lua/utils.c

```diff
@@ -59,6 +69,64 @@ uint32_t CTID_UUID;
 uint32_t CTID_DATETIME = 0;
 uint32_t CTID_INTERVAL = 0;
 
+/**
+ * Lua custom memory allocation function. It extends the original one with
+ * a memory counter and a limit check.
+ */
+static void *
+luaT_alloc(void *ud, void *ptr, size_t osize, size_t nsize)
+{
+	alloc.allocated += nsize - osize;
+	if (alloc.allocated > alloc.limit) {
+		diag_set(OutOfMemory, nsize, "luaT_alloc", "lua objects");
+		diag_log();
+		return NULL;
+	}
+
+	void *result = alloc.orig_alloc(ud, ptr, osize, nsize);
+	return result;
+}
+
+struct lua_State *
+luaT_newstate(void)
+{
+	struct lua_State *L = luaL_newstate();
+	if (L == NULL) {
+		panic("failed to initialize Lua");
+	}
+
+	/*
+	 * lua_gc(L, LUA_GC_COUNT) returns the allocated memory in KBs
+	 * lua_gc(L, LUA_GC_COUNTB) returns the reminder
+	 */
+	alloc.allocated = lua_gc(L, LUA_GCCOUNT, 0) * 1024ll +
```

Comment:
Why do we need `ll` suffix here?

Reply:
`lua_gc` returns an `int` number. Without `ll` an implicit conversion into a `long long` isn't performed. An overflow is possible.